src/iop/lut3d.c

```diff
@@ -0,0 +1,1023 @@
+       transform = FALSE;
+    }
+    size_t sizes[] = { ROUNDUPWD(width), ROUNDUPHT(height), 1 };
+    if (transform)
+      dt_opencl_set_kernel_arg(devid, kernel, 0, sizeof(cl_mem), (void *)&dev_out);
+    else
+      dt_opencl_set_kernel_arg(devid, kernel, 0, sizeof(cl_mem), (void *)&dev_in);
+    dt_opencl_set_kernel_arg(devid, kernel, 1, sizeof(cl_mem), (void *)&dev_out);
+    dt_opencl_set_kernel_arg(devid, kernel, 2, sizeof(int), (void *)&width);
+    dt_opencl_set_kernel_arg(devid, kernel, 3, sizeof(int), (void *)&height);
+    dt_opencl_set_kernel_arg(devid, kernel, 4, sizeof(cl_mem), (void *)&clut_cl);
+    dt_opencl_set_kernel_arg(devid, kernel, 5, sizeof(int), (void *)&level);
+    err = dt_opencl_enqueue_kernel_2d(devid, kernel, sizes);
+    if (transform)
+      dt_ioppr_transform_image_colorspace_rgb_cl(devid, dev_out, dev_out, width, height, lut_profile, work_profile, "LUT Space to linrec2020");
+  }
+  else  // no clut, do nothing
+  {
+    err = CL_INVALID_VALUE;
+  }
+  if(err != CL_SUCCESS)
+  {
+    fprintf(stderr, "[lut3d process_cl] error %i enqueue kernel\n", err);
+    goto cleanup;
+  }
+
+cleanup:
+  if(clut_cl) dt_opencl_release_mem_object(clut_cl);
+
+  if(err != CL_SUCCESS) dt_print(DT_DEBUG_OPENCL, "[opencl_lut3d] couldn't enqueue kernel! %d\n", err);
+  return (err == CL_SUCCESS) ? TRUE : FALSE;
+}
+#endif
+
+void process(struct dt_iop_module_t *self, dt_dev_pixelpipe_iop_t *piece, const void *const ibuf, void *const obuf,
+             const dt_iop_roi_t *const roi_in, const dt_iop_roi_t *const roi_out)
+{
+  dt_iop_lut3d_data_t *d = (dt_iop_lut3d_data_t *)piece->data;
+  const int ch = piece->colors;
+  const float *const clut = (float *)d->clut;
+  const uint8_t level = d->level;
+  const dt_interpolation_worker interpolation_worker
+    = (d->params.interpolation == DT_IOP_TETRAHEDRAL) ? correct_pixel_tetrahedral
+    : (d->params.interpolation == DT_IOP_TRILINEAR) ? correct_pixel_trilinear
+    : correct_pixel_pyramid;
+  const int colorspace
+    = (d->params.colorspace == DT_IOP_SRGB) ? DT_COLORSPACE_SRGB
+    : (d->params.colorspace == DT_IOP_REC709) ? DT_COLORSPACE_REC709
+    : (d->params.colorspace == DT_IOP_LIN_REC709) ? DT_COLORSPACE_LIN_REC709
+    : DT_COLORSPACE_LIN_REC2020;
+  const dt_iop_order_iccprofile_info_t *const lut_profile
+    = (colorspace != DT_COLORSPACE_LIN_REC2020) ? dt_ioppr_add_profile_info_to_list(self->dev, colorspace, "", INTENT_PERCEPTUAL) : NULL;
+  const dt_iop_order_iccprofile_info_t *const work_profile
+    = dt_ioppr_get_pipe_work_profile_info(piece->pipe);
+  const gboolean transform = (work_profile != NULL && lut_profile != NULL) ? TRUE : FALSE;
+  if (clut)
+  {
+    if (transform)
+    {
+      dt_ioppr_transform_image_colorspace_rgb(ibuf, obuf, roi_in->width, roi_in->height, work_profile, lut_profile, "linrec2020 to LUT Space");
+#ifdef _OPENMP
+#pragma omp parallel for default(none) schedule(static)
+#endif
+      for(int j = 0; j < roi_out->height; j++)

```

Comment:
You can easily collapse the 2 loops to help the autovectorization and parallel loops splitting:
```C
for(size_t k; k < roi_out->width * roi_out->height * ch; k += ch)
    float *const out = ((float *const)obuf) + k;
    interpolation_worker(out, out, clut, level);
```

Reply:
On another note, it might be good to use 2 different buffers, such that `interpolation(in, out, clut, level)`, to make better use of the cache and speed it up a bit (to be tested). This way, you can declare `void correct_pixel_tetrahedral(float *const restrict input, float *const restrict output, const float *const restrict clut, const uint8_t level)`

Reply:
Loop done. 
To test the efficiency of the second one I would need to measure the execution time. What I've made so far (pick up system time at the beginning and end of process varies a lot) is not accurate. Is there an existing protocol I could follow ?

Reply:
One basic test would be to launch dt with `-d perf` option and compare the timing. For the loop, I can already tell you it will be better with the collapsed version because OpenMP is able to evenly divide the array between the threads, whereas with 2 nested loops, the last chunk is likely to be smaller than the others.

Reply:
I thought it was just a matter of copying the pixel data in another place before calling interpolation. But your suggestion is to copy the full image instead, is that correct ?

Reply:
Your interpolation works inplace, meaning it takes the `out` buffer as an input and overwrite it (so *input = *output). When you want to (auto)vectorize the code, it's usually better to use read-only and write-only buffers, so maybe copy the `obuf` somewhere else, then *input ≠ *output and the compiler should be able to perform further optimizations.

Reply:
... inside optimization routine `input` is not used a lot compared to other calculations. Not sure it could improve really. Should be measured but I haven't time for this currently. However here is the new code if I understand properly (I've also put the `restrict` directive on the routines' header):
```
      for(int i = 0; i < width * height * ch; i+=ch)
      {
        float input[4];
        float *const out = ((float *const)obuf) + i;
        for (int j = 0; j < ch; j++) input[j] = out[j];
        interpolation_worker((float *const)&input, out, clut, level);
      }
```
Is that in line with your suggestion ? If ok, I'll push the new code.

Reply:
Above code is pushed.